### PR TITLE
Windows build

### LINKED ISSUE #107: "$(uname)" was unexpected at this time.
I followed the instructions of msys2.

> ```
> # update the system
> # in case of errors, wait for the update to complete
> # then close and open again MingW shell
> pacman -Syyu --noconfirm
> 
> # install git, gtk3 and extra dependencie
> pacman -S --needed --noconfirm git mingw-w64-$(uname -m)-{gtk3,gobject-introspection,pkg-config}
> 
> # where to put the repository clone?
> # pick your flder or use ~/oss (Open Source Software)
> mkdir -p ~/oss/
> cd ~/oss
> 
> # clone node-gtk there
> git clone https://github.com/romgrk/node-gtk
> cd node-gtk
> 
> # first run might take a while
> GYP_MSVS_VERSION=2015 npm install
> ```

And in the last step, I got the following errors.

```
Liu.D.H@DESKTOP-KA8TQF4 MINGW32 ~/oss/node-gtk
# GYP_MSVS_VERSION=2017 npm install

> node-gtk@0.2.2 install C:\msys32\home\Liu.D.H\oss\node-gtk
> if [ "$(uname)" = "Darwin" ] && [ "$(which brew)" != "" ]; then export PKG_CONFIG_PATH=$(brew --prefix libffi)/lib/pkgconfig; fi; node-pre-gyp install --fallback-to-build

"$(uname)" was unexpected at this time.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-gtk@0.2.2 install: `if [ "$(uname)" = "Darwin" ] && [ "$(which brew)" != "" ]; then export PKG_CONFIG_PATH=$(brew --prefix libffi)/lib/pkgconfig; fi; node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the node-gtk@0.2.2 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Liu.D.H\.npm\_logs\2019-04-05T15_31_25_342Z-debug.log

Liu.D.H@DESKTOP-KA8TQF4 MINGW32 ~/oss/node-gtk
#
```

I have installed the nodejs in my msys2 environment and visual studio 2017 build tools outside of msys2. I also tried `yarn` or execute this command in my cmd. But none of these methods worked.

```
Liu.D.H@DESKTOP-KA8TQF4 MINGW32 ~/oss/node-gtk
# yarn
yarn install v1.15.2
info No lockfile found.
warning package-lock.json found. Your project contains lock files generated by tools other than Yarn. It is advised not to mix package managers in order to avoid resolution
 inconsistencies caused by unsynchronized lock files. To clear this warning, remove package-lock.json.
[1/4] Resolving packages...
warning node-pre-gyp-github > github@0.2.4: 'github' has been renamed to '@octokit/rest' (https://git.io/vNB11)
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...

success Saved lockfile.
$ if [ "$(uname)" = "Darwin" ] && [ "$(which brew)" != "" ]; then export PKG_CONFIG_PATH=$(brew --prefix libffi)/lib/pkgconfig; fi; node-pre-gyp install --fallback-to-build
"$(uname)" was unexpected at this time.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

Liu.D.H@DESKTOP-KA8TQF4 MINGW32 ~/oss/node-gtk
#
```

### PR DESCRIPTION
**Pull request by ten0s**
This PR makes it possible to build and run on Windows using Node.js 12, see GitHub workflow for Windows.

At this point the next examples run successfully:

examples/hello-gtk.js
examples/drawing-area.js
examples/builder-auto-connect-signals.js
examples/builder.js
examples/entry.js

Closes #107

### CODE REVIEW

**Comment by ten0s** [comment]
Screenshot_2022-06-09_16-18-21

**Comment by romgrk** [comment]
Amazing work :open_mouth: I have limited time these days but ping me when it's ready, I'll do a quick review and as long as the tests are passing for the existing OSes I'm ok with merging it.

**Comment by ten0s** [comment]
@romgrk At this point all the tests pass on Linux and the majority of tests pass on Windows.

The two skipped tests callback.js and signal__non-introspected.js (both Gst) pass locally on my machine, but seems to crash in CI. Since Gst doesn't interest me I'm not going to look into them.

Also I disabled one test inside conversion__array.js introduced in https://github.com/romgrk/node-gtk/pull/313, since it crashes somewhere inside node. See comment in code. 

MacOS build still fails https://github.com/romgrk/node-gtk/pull/326, so no idea if my changes somehow affected it.

**Comment by romgrk** [comment]
I've started the test suite, let's see what passes.

Overall LGTM, I'm ok merging as it is. Some windows support is better than no windows support.

One thing though, for the `cairo.cc` changes, those files are actually autogenerated, can you update the corresponding generator?
https://github.com/romgrk/node-gtk/blob/ab813bb45e2801fd7ffd9dc9b441a06d520c1143/src/modules/cairo/generator.js#L524-L536

**Comment by ten0s** [comment]
Will do.

Seems like I found the cause of 
https://github.com/romgrk/node-gtk/pull/327/commits/a1e0f72660d4d493ac6025fcf83ce91be7e40f35#diff-2ecb150c1fd88f2ab8aaf603ccee701e18d113bdb931ffc3aa4eb528243a3a38R1347-R1349

Will change, test and give you know.

**Comment by ten0s** [comment]
@romgrk Done

File: README.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ Node-Gtk is a [gobject-introspection](https://gi.readthedocs.io/en/latest) libra
 use any introspected library, such as Gtk+, usable.  It is similar in essence to [GJS](https://wiki.gnome.org/action/show/Projects/Gjs) or [PyGObject](https://pygobject.readthedocs.io). Please note this project is currently in a _beta_ state and is being developed. Any contributors willing to help
 will be welcomed.
 
-Supported Node.js versions: **12**, **14**, **15**, **16** (other versions should work but are untested)  
+Supported Node.js versions: **12**, **14**, **15**, **16** (other versions should work but are untested)
```
Comment (romgrk) [inline README.md:22]: Can you undo this one, the double-space is to insert a newline here.

File: src/function.cc [hunk 2]
```diff
@@ -325,7 +324,7 @@ Local<Value> FunctionCall (
      * and for error, if it can throw
      */
 
-    GIArgument total_arg_values[func->n_total_args];
+    GIArgument *total_arg_values = new GIArgument[func->n_total_args]();
```
Comment (romgrk) [inline src/function.cc:327]: Oh right we were using [this form](https://github.com/romgrk/node-gtk/blob/56335f82387011ecb590a8bffe88e35c7cb27d02/src/closure.cc#L50-L54) to deal with this. Can we do the same here? `FunctionCall` runs on every function call (who would have guessed?) so I'd rather avoid the allocation if possible.

**Review by romgrk — COMMENTED** [review]
A few comments, but LGTM. If you're not sure about the macos stuff try to leave it untouched.

File: README.md (hunk 1 above)
Comment (ten0s) [reply README.md]: It took me a while to understand why it should be reverted)))

**Comment by ten0s** [comment]
@romgrk Done

**Comment by ten0s** [comment]
Great! Hopefully the project will start gaining more traction!

**Comment by romgrk** [comment]
Yeah, my dream as always been to have a clean & simple native alternative to electron. `node-qt` is probably better placed than this project, but the downside is they use Qt which is ugly by default.

**Pull request merged**